code-review/class.HTTPUtilities.php

```diff
@@ -0,0 +1,864 @@
+        $this->_auditor = $ESAPI::getAuditor('DefaultHTTPUtilities');
+	$this->_validator = $ESAPI::getValidator();
+
+    }
+
+
+    /**
+     * Adds the CSRF token from the current session to the supplied URL for the
+     * purposes of preventing CSRF attacks. This method should be used on all URLs
+     * to be put into all links and forms the application generates.
+     *
+     * @param string $href the URL to which the CSRF token will be appended.
+     *
+     * @return string URL with the CSRF token parameter appended to it.
+     */
+    public function addCSRFToken($href)
+    {
+        if (! is_string($href) || empty($href)) {
+            throw new InvalidArgumentException(
+                'addCSRFToken expects string $href.'
+            );
+        }
+        if (! isset($_SESSION)) {
+            return $href;
+        }
+        
+        $token = $this->getCSRFToken();
+        if ($token === null) {
+            return $href;
+        }
+        
+        if (strpos($href, '?') === false) {
+            $href .= '?' . $token;
+        } else {
+            $href .= '&' . $token;
+        }
+        
+        return $href;
+    }
+
+
+    /**
+     * Returns the CSRF token from the current session. If there is no current
+     * session then null is returned. If the CSRF Token is not present in the
+     * session it will be created.
+     *
+     * @return string|null CSRF token for the current session or
+     *                     null.
+     */
+    public function getCSRFToken()
+    {
+        if (! isset($_SESSION)) {
+            return null;
+        }
+        
+        if (   ! array_key_exists('ESAPI', $_SESSION)
+            || ! array_key_exists('HTTPUtilities', $_SESSION['ESAPI'])
+            || ! array_key_exists('CSRFToken', $_SESSION['ESAPI']['HTTPUtilities'])
+        ) {
+            $this->setCSRFToken();
+        }
+        
+        return $_SESSION['ESAPI']['HTTPUtilities']['CSRFToken'];
+    }
+
+
+    /**
+     * Searches the GET and POST parameters in a request for the CSRF token stored
+     * in the current session and throws an IntrusionException if it is missing.
+     *
+     * @param SafeRequest $request A request object.
+     *
+     * @return null
+     *
+     * @throws IntrusionException if the CSRF token is missing or incorrect.
+     */
+    public function verifyCSRFToken($request)
+    {
+        if ($request instanceof SafeRequest == false) {
+            throw new InvalidArgumentException(
+                'verifyCSRFToken expects an instance of SafeRequest.'
+            );
+        }
+
+        if ($request->getParameter($this->getCSRFToken()) === null) {
+            throw new IntrusionException(
+                'Authentication failed.',
+                'Possibly forged HTTP request without proper CSRF token detected.'
+            );
+        }
+        
+    }
+
+
+    /**
+     * Sets the CSRF Token for the current session.  If the session has not been
```

Comment:
### Pažeidžiamos taisyklės
- [ ] Komentarai turėtų būti parašyti be gramatinių klaidų
### Komentarai
Nereikia palikti dvigubo tarpo:
```suggestion
     * Sets the CSRF Token for the current session. If the session has not been
```